daemon/logger/jsonfilelog/jsonfilelog.go

```diff
@@ -124,72 +124,85 @@ func (l *JSONFileLogger) Log(msg *logger.Message) error {
 }
 
 func writeLog(l *JSONFileLogger) (int64, error) {
+
+	if err := checkCapacityAndRotate(l); err != nil {
+		return -1, err
+	}
+
+	i, err := l.buf.WriteTo(l.f)
+	if err != nil {
+		l.buf.Reset()
+	}
+
+	return i, err
+}
+
+func checkCapacityAndRotate(l *JSONFileLogger) error {
 	if l.capacity == -1 {
-		return writeToBuf(l)
+		return nil
 	}
+
 	meta, err := l.f.Stat()
 	if err != nil {
-		return -1, err
+		return err
 	}
 	if meta.Size() >= l.capacity {
 		name := l.f.Name()
 		if err := l.f.Close(); err != nil {
-			return -1, err
+			return err
 		}
-		if err := rotate(name, l.n); err != nil {
-			return -1, err
+		if err := rotate(name, l.maxFiles); err != nil {
+			return err
 		}
 		file, err := os.OpenFile(name, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0666)
 		if err != nil {
-			return -1, err
+			return err
 		}
 		l.f = file
 		l.notifyRotate.Publish(struct{}{})
 	}
-	return writeToBuf(l)
-}
 
-func writeToBuf(l *JSONFileLogger) (int64, error) {
-	i, err := l.buf.WriteTo(l.f)
-	if err != nil {
-		l.buf = bytes.NewBuffer(nil)
-	}
-	return i, err
+	return nil
 }
 
-func rotate(name string, n int) error {
-	if n < 2 {
+func rotate(name string, maxFiles int) error {
+	if maxFiles < 2 {
 		return nil
 	}
-	for i := n - 1; i > 1; i-- {
-		oldFile := name + "." + strconv.Itoa(i)
-		replacingFile := name + "." + strconv.Itoa(i-1)
-		if err := backup(oldFile, replacingFile); err != nil {
+	for i := maxFiles - 1; i > 1; i-- {
+		toPath := name + "." + strconv.Itoa(i)
+		fromPath := name + "." + strconv.Itoa(i-1)
+
+		if _, err := os.Stat(fromPath); os.IsNotExist(err) {
+			continue
```

Comment:
I think it would be better to adjust the `backup` function to not rename the file if the `fromPath` doesn't exist. Then this goes away.